File: app/src/main/java/com/example/mainactivity/ui/chat/ChatTimeFormatters.kt
package com.example.mainactivity.ui.chat

import java.time.Instant
import java.time.OffsetDateTime
import java.time.ZoneId
import java.time.format.DateTimeFormatter
import java.util.Calendar
import java.util.Locale

private const val MILLIS_PER_MINUTE = 60_000L
private const val MILLIS_PER_HOUR = 3_600_000L
private const val MILLIS_PER_DAY = 86_400_000L
private const val MINUTES_PER_HOUR = 60
private const val HOURS_PER_DAY = 24
private const val DAYS_PER_WEEK = 7
private const val PRESENCE_ACTIVE_NOW_MINUTES = 2
private const val MESSAGE_GROUP_GAP_MILLIS = 10 * MILLIS_PER_MINUTE

private fun dayOfWeekShort(epochMs: Long): String {
    val cal = Calendar.getInstance()
    cal.timeInMillis = epochMs
    return cal.getDisplayName(Calendar.DAY_OF_WEEK, Calendar.SHORT, Locale.getDefault()) ?: ""
}

private fun monthDayShort(epochMs: Long): String {
    val cal = Calendar.getInstance()
    cal.timeInMillis = epochMs
    val month = cal.getDisplayName(Calendar.MONTH, Calendar.SHORT, Locale.getDefault()) ?: ""
    return "$month ${cal.get(Calendar.DAY_OF_MONTH)}"
}

/** Short relative time label for conversation list preview (e.g. "2m ago", "Yesterday", "Mon"). */
internal fun relativeTime(isoString: String): String =
    runCatching {
        val instant = Instant.parse(isoString)
        val diffMs = System.currentTimeMillis() - instant.toEpochMilli()
        val diffMin = diffMs / MILLIS_PER_MINUTE
        val diffH = diffMs / MILLIS_PER_HOUR
        val diffD = diffMs / MILLIS_PER_DAY
        when {
            diffMin < 1 -> "now"
            diffMin < MINUTES_PER_HOUR -> "${diffMin}m ago"
            diffH < HOURS_PER_DAY -> "${diffH}h ago"
            diffD == 1L -> "Yesterday"
            diffD < DAYS_PER_WEEK -> dayOfWeekShort(instant.toEpochMilli())
            else -> monthDayShort(instant.toEpochMilli())
        }
    }.getOrDefault("")

/** Compact time label for message timestamps (e.g. "2:30 PM", "Yesterday 2:30 PM", "Mon 2:30 PM"). */
internal fun messageTimeLabel(isoString: String): String =
    runCatching {
        val instant = Instant.parse(isoString)
        val diffD = (System.currentTimeMillis() - instant.toEpochMilli()) / MILLIS_PER_DAY
        val odt = OffsetDateTime.ofInstant(instant, ZoneId.systemDefault())
        val timePart = odt.format(DateTimeFormatter.ofPattern("h:mm a"))
        when {
            diffD == 0L -> timePart
            diffD == 1L -> "Yesterday $timePart"
            diffD < DAYS_PER_WEEK -> "${dayOfWeekShort(instant.toEpochMilli())} $timePart"
            else -> "${monthDayShort(instant.toEpochMilli())} $timePart"
        }
    }.getOrDefault("")

/** Chat-header presence: "Active now" within 2 minutes, else "Active {relative}", or null. */
internal fun presenceLabel(lastActiveIso: String?): String? {
    if (lastActiveIso == null) return null
    return runCatching {
        val instant = Instant.parse(lastActiveIso)
        val diffMin = (System.currentTimeMillis() - instant.toEpochMilli()) / MILLIS_PER_MINUTE
        if (diffMin < PRESENCE_ACTIVE_NOW_MINUTES) "Active now" else "Active ${relativeTime(lastActiveIso)}"
    }.getOrNull()
}

/** True if another participant's last-read time [otherLastRead] is at or after [sentAt],
 *  i.e. they've seen the message. */
internal fun messageSeen(
    otherLastRead: String?,
    sentAt: String,
): Boolean {
    if (otherLastRead == null) return false
    return runCatching {
        !Instant.parse(otherLastRead).isBefore(Instant.parse(sentAt))
    }.getOrDefault(false)
}

/** Returns true if the gap between two ISO timestamps exceeds 10 minutes. */
internal fun gapExceedsTenMinutes(
    earlierIso: String,
    laterIso: String,
): Boolean =
    runCatching {
        val earlier = Instant.parse(earlierIso).toEpochMilli()
        val later = Instant.parse(laterIso).toEpochMilli()
        (later - earlier) > MESSAGE_GROUP_GAP_MILLIS
    }.getOrDefault(false)

File: ios/FamilyApp/Features/Chat/ChatTimeFormatters.swift
// Chat time/receipt helpers — the iOS twin of ChatTimeFormatters.kt. Pure and
// unit-tested; `now` is injectable for tests.
import Foundation

private let minuteMs: Int64 = 60000
private let hourMs: Int64 = 3600000
private let dayMs: Int64 = 86400000
private let presenceActiveNowMinutes: Int64 = 2
private let messageGroupGapMs: Int64 = 10 * minuteMs

func parseInstantMs(_ isoString: String) -> Int64? {
    let formatter = ISO8601DateFormatter()
    formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
    if let date = formatter.date(from: isoString) {
        return Int64(date.timeIntervalSince1970 * 1000)
    }
    formatter.formatOptions = [.withInternetDateTime]
    guard let date = formatter.date(from: isoString) else { return nil }
    return Int64(date.timeIntervalSince1970 * 1000)
}

private func dayOfWeekShort(_ epochMs: Int64) -> String {
    let formatter = DateFormatter()
    formatter.dateFormat = "EEE"
    return formatter.string(from: Date(timeIntervalSince1970: Double(epochMs) / 1000))
}

private func monthDayShort(_ epochMs: Int64) -> String {
    let formatter = DateFormatter()
    formatter.dateFormat = "MMM d"
    return formatter.string(from: Date(timeIntervalSince1970: Double(epochMs) / 1000))
}

private func timePart(_ epochMs: Int64) -> String {
    let formatter = DateFormatter()
    formatter.dateFormat = "h:mm a"
    formatter.locale = Locale(identifier: "en_US_POSIX")
    return formatter.string(from: Date(timeIntervalSince1970: Double(epochMs) / 1000))
}

/// Short relative label for the conversation list (e.g. "2m ago", "Yesterday", "Mon").
func relativeTime(_ isoString: String, nowMs: Int64 = Int64(Date().timeIntervalSince1970 * 1000)) -> String {
    guard let instant = parseInstantMs(isoString) else { return "" }
    let diffMs = nowMs - instant
    let diffMin = diffMs / minuteMs
    let diffH = diffMs / hourMs
    let diffD = diffMs / dayMs
    switch true {
    case diffMin < 1: return "now"
    case diffMin < 60: return "\(diffMin)m ago"
    case diffH < 24: return "\(diffH)h ago"
    case diffD == 1: return "Yesterday"
    case diffD < 7: return dayOfWeekShort(instant)
    default: return monthDayShort(instant)
    }
}

/// Compact message timestamp (e.g. "2:30 PM", "Yesterday 2:30 PM", "Mon 2:30 PM").
func messageTimeLabel(_ isoString: String, nowMs: Int64 = Int64(Date().timeIntervalSince1970 * 1000)) -> String {
    guard let instant = parseInstantMs(isoString) else { return "" }
    let diffD = (nowMs - instant) / dayMs
    let time = timePart(instant)
    switch true {
    case diffD == 0: return time
    case diffD == 1: return "Yesterday \(time)"
    case diffD < 7: return "\(dayOfWeekShort(instant)) \(time)"
    default: return "\(monthDayShort(instant)) \(time)"
    }
}

/// Chat-header presence: "Active now" within 2 minutes, else "Active {relative}".
func presenceLabel(_ lastActiveIso: String?, nowMs: Int64 = Int64(Date().timeIntervalSince1970 * 1000)) -> String? {
    guard let lastActiveIso, let instant = parseInstantMs(lastActiveIso) else { return nil }
    let diffMin = (nowMs - instant) / minuteMs
    return diffMin < presenceActiveNowMinutes
        ? "Active now"
        : "Active \(relativeTime(lastActiveIso, nowMs: nowMs))"
}

/// True when another participant's last_read_at is at or after the message's sent_at.
func messageSeen(otherLastRead: String?, sentAt: String) -> Bool {
    guard let otherLastRead,
          let read = parseInstantMs(otherLastRead),
          let sent = parseInstantMs(sentAt) else { return false }
    return read >= sent
}

/// True if the gap between two ISO timestamps exceeds 10 minutes (time separators).
func gapExceedsTenMinutes(earlierIso: String, laterIso: String) -> Bool {
    guard let earlier = parseInstantMs(earlierIso), let later = parseInstantMs(laterIso) else {
        return false
    }
    return later - earlier > messageGroupGapMs
}

// MARK: - Display-name / preview resolution (mirrors ConversationRow logic)

func conversationDisplayName(
    conversation: ConversationModel,
    participants: [UserModel],
    currentUserId: String
) -> String {
    let isOneOnOne = participants.count == 2 || conversation.userTo != nil
    let other = participants.first { $0.id != currentUserId }
    if isOneOnOne {
        if let name = other?.name, !name.isEmpty { return name }
        return conversation.name.isEmpty ? "Chat" : conversation.name
    }
    if !conversation.name.isEmpty { return conversation.name }
    let names = participants
        .filter { $0.id != currentUserId }
        .prefix(3)
        .compactMap { $0.name.split(separator: " ").first.map(String.init) }
        .joined(separator: ", ")
    return names.isEmpty ? "Group chat" : names
}

func conversationPreviewText(lastMessage: MessageModel?, lastSenderName: String?) -> String {
    guard let lastMessage else { return "No messages yet" }
    let prefix = lastSenderName.map { "\($0): " } ?? ""
    switch lastMessage.messageType {
    case "image": return "\(prefix)Photo"
    case "voice": return "\(prefix)Voice message"
    default: return "\(prefix)\(lastMessage.text)"
    }
}
